bl31/aarch64/bl31_entrypoint.S

```diff
@@ -42,6 +43,39 @@
 	 */
 
 func bl31_entrypoint
+#ifdef BL31_ENABLE_RELOCATION
+	/*
+	 * Below code snippet is used for relocating BL3-1.
+	 *
+	 *The BL3-1 relocation ensures that BL3-1 only runs from compiled
+	 *execution address. If previous boot-loader stage loads BL3-1 to
+	 *an address different from compiled execution address then BL3-1
+	 *relocation will copy BL3-1 to desired execution address and continue
+	 *from there.
+	 */
+
+	adr	x10, bl31_entrypoint
+	ldr	x11, __reloc_start
+	ldr	x12, __reloc_end
+	cmp	x10, x11
+	b.eq	_reloc_done
+1:	ldp	x13, x14, [x10], #16
+	stp	x13, x14, [x11], #16
```

Comment:
As discussed in https://github.com/ARM-software/tf-issues/issues/382, implementation needs to check for a destructive overlapping copy "in the case that the end of the loaded image region overlaps the start of the run region". Suggest using ASM_ASSERT for this.